dev-tool/src/call_data/generate.rs

```diff
@@ -0,0 +1,492 @@
+#![cfg(not(tarpaulin))]
+use std::{
+    cmp::min,
+    collections::{HashMap, HashSet},
+    fmt,
+    fs::File,
+    io::BufReader,
+    path::{Path, PathBuf},
+};
+
+use anyhow::{bail, Error};
+use rand::{
+    prelude::{IteratorRandom, SliceRandom, ThreadRng},
+    thread_rng,
+    Rng,
+};
+use serde::Deserialize;
+use structopt::StructOpt;
+use uuid::Uuid;
+use xayn_ai::{
+    DayOfWeek,
+    Document,
+    DocumentHistory,
+    DocumentId,
+    QueryId,
+    Relevance,
+    RerankMode,
+    SessionId,
+    UserAction,
+    UserFeedback,
+};
+
+use crate::exit_code::NO_ERROR;
+
+use super::CallData;
+
+/// Generate call data for the example.
+#[derive(StructOpt, Debug)]
+pub struct GenerateCallDataCmd {
+    /// Json file to write the generate call data to.
+    #[structopt(short, long)]
+    out: PathBuf,
+
+    /// File with snippets to use.
+    #[structopt(long)]
+    snippets: PathBuf,
+
+    /// Number of documents in current query.
+    ///
+    /// I.e. the number of documents the result of query (we
+    /// pretend to run) in the example app has.
+    #[structopt(short = "q", long)]
+    number_of_current_documents: usize,
+
+    /// Number of queries in the users history.
+    ///
+    /// Each query will randomly have between 5 and 30 documents.
+    #[structopt(short = "h", long)]
+    number_of_historic_queries: usize,
+
+    /// Extend domains and urls.
+    ///
+    /// Normally urls will be generated using following
+    /// schema: `url://dom_{domain_id}/{url_id}`.
+    ///
+    /// This uses more or less incremental ids but this
+    /// means that domains and urls are unrealistic short.
+    ///
+    /// Using this option will make the domains and URLs
+    /// longer. You can use it multiple times.
+    #[structopt(short = "l", parse(from_occurrences))]
+    lengthen_urls: usize,
+}
+
+const MIN_QUERY_NR_DOCS: usize = 5;
+const MAX_QUERY_NR_DOCS: usize = 30;
+
+impl GenerateCallDataCmd {
+    pub fn run(self) -> Result<i32, Error> {
+        let GenerateCallDataCmd {
+            out,
+            snippets,
+            number_of_current_documents,
+            number_of_historic_queries,
+            lengthen_urls,
+        } = self;
+
+        if !(MIN_QUERY_NR_DOCS..=MAX_QUERY_NR_DOCS).contains(&number_of_current_documents) {
```

Comment:
Why do we have this? I think we can have only a lower bound of 1 but if in the future we need 100 we can just input 100 and have 30 as default.

Reply:
Sure, I mainly used it to constraint the random generated history. I can change this to be 1+.